### ISSUE TITLE
[Feature request] download location 

### ISSUE THREAD

**Issue by karlprieb** [CappielloAntonio/tempo#13]
Hey 👋 
I found Tempo and I’ve been using it as my primary music player. It’s great! Fast and beautiful.

One thing that I miss is to be able to select where it will save downloads. I use it on my android DAP that has very limited internal storage and would be good if I can save the downloads on my sd card instead.

**Comment by CappielloAntonio** [comment]
Hi, it's been a long time. These days I have added the ability to select external storage as the default storage for downloads. It's a new option within the settings, visible only if the app detects an external memory.

I was mainly slowed down by the impossibility of testing the functionality with my own physical device, so I relied on the Android Studio emulator (a love/hate relationship).

Back to functionality:
It seems to work, the download is basically handled by Media3 which uses the default app folder. Due to the increasingly stringent storage access policies adopted by Google, this allows me not to ask the user for permission to write the download.

There are three limitations:
- I can only use one storage at a time. Therefore, I cannot have files downloaded to both storages at the same time (or better, it is possible, but I can only read from one location).
- I can't move downloads from one storage to another, moving from one place to another the user must accept losing all the previously saved downloads.
- The change will take effect the first time the app is restarted, a manual operation that the user have to do.

Bottom line, would you be willing to try this new version and tell me what you think? 
I'd be interested to know mainly if it works or there is strange behavior. 

I am attaching to this comment a zip containing the new apk for anyone who wants to try it. Let me know!

[app-tempo-debug.zip](https://github.com/CappielloAntonio/tempo/files/12264937/app-tempo-debug.zip)

**Comment by karlprieb** [comment]
Hey @CappielloAntonio 👋 
Thank you for implementing this!

I just tested and works as you described!

Once I change the setting to external storage the app display this warning modal
![Screenshot_20230805-163217](https://github.com/CappielloAntonio/tempo/assets/9200155/99791702-df00-421f-ba94-36afce2b323f)

I downloaded an album
![Screenshot_20230805-163509](https://github.com/CappielloAntonio/tempo/assets/9200155/2819563d-8e8d-4968-9333-f5662302d7ab)

and checked the app folder in the sdcard
![Screenshot_20230805-163555](https://github.com/CappielloAntonio/tempo/assets/9200155/a5358963-7b15-4946-8e2a-c608e56c05f2)

After that I played the album with the device in offline mode and it worked just fine :)
![Screenshot_20230805-163647 2](https://github.com/CappielloAntonio/tempo/assets/9200155/a778be88-0338-4c18-949d-1cb152106444)

I changed the download storage to bathe internal and all downloads were gone as you said :+1:
One thing that I noticed is that the downloaded files on sdcard was still there.

knowing that I have two suggestions:
- when the user change the storage setting you can doce close the app to make sure it was restarted
- as you will loose the reference for the files when changing storage you could delete them when changing the storage

thanks again :)

**Comment by CappielloAntonio** [comment]
Hi @karlprieb, thanks a lot for the test.
To respond to your comments:
- I guess there's not much I can do to force restart the app, I could destroy and recreate the single activity, but I'm afraid it wouldn't have the desired effect for the download destination folder switch.
- That the downloads stay in the destination folder should definitely not happen, I'll do some more testing.

**Comment by karlprieb** [comment]
Sure :)
Update me if you need me to test it again

**Comment by CappielloAntonio** [comment]
Hi @karlprieb, here we go again.

I have a few questions to ask:
- is it possible that when storage is changed and after restarting, what is left in memory are just empty folders?
- would it be possible for you to test this functionality in a clean environment, deleting all your downloads and starting from scratch?

I'm doing several tests locally (on an emulator as I don't yet have a physical device that allows me to test) and when storage is changed or storage is forced to be deleted, all the elements in the db are effectively deleted. What I do is read the db, read the downloads there, and for each download I delete the file in memory.

My doubt is that you somehow have items in memory that aren't saved in the db. This could be caused by some carelessness of mine in one of the updates or I don't know, it seems really strange to me.

**Comment by karlprieb** [comment]
I can test it again from a clean state, like I did on my previous test, tomorrow :)

> is it possible that when storage is changed and after restarting, what is left in memory are just empty folders?
from what I remember the downloaded files were still there, but I’ll test it again to make sure.

> My doubt is that you somehow have items in memory that aren't saved in the db. This could be caused by some carelessness of mine in one of the updates or I don't know, it seems really strange to me.
Maybe 🤔 maybe the download failed somehow and you didn’t inserted it in the db, but some chunks were saved in the file system. I don’t know :)
If I can reproduce it tomorrow I’ll check the code tomorrow to understand how you implemented it and maybe I can catch something.

**Comment by karlprieb** [comment]
Sorry for not providing feedback :)
I was able to test now with the latest build and it works as expected!
